### PR TITLE
colexec: ensure hash table emit first tuple

### PR DESCRIPTION
**Pull request by Azhng**
colexec: ensure hashtable emits first distinct tuple when built
in distinct mode.

Previously, hashtable when built in distinct mode, hashtable
does not guarentee the order of the output.
This causes the vectorize unordered distinct to behave differently
from the unordered distinct in the row engine.
Now hashtable when built in distinct mode, it is guaranteed to
emit the first distinct tuple.

Release justification: bug fixes and low-risk updates to new
functionality since we are not turning unordered distinct on
by default

Release note: None

### CODE REVIEW

**Review by yuzefovich — CHANGES_REQUESTED** [review]
You should also update `TestDistinctAgainstProcessor` test and remove `colForEqCheck` thing entirely so that we compare the whole rows always.

> Reviewed 2 of 2 files at r1, 11 of 11 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M20rplk3Y4trwKf1-tl:bn18wzu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[pkg/sql/colexec/distinct_test.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20ofxiD43Lk0vNRYMi:-M20ofxiD43Lk0vNRYMj:be48284) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/distinct_test.go#L133)):*
> ```Go
> 			colTypes:     []coltypes.T{coltypes.Int64, coltypes.Bytes},
> 			tuples: tuples{
> 				{1, "1"},
> ```

nit: maybe make the order here more randomized? Also I think you should add `nil` in the grouping columns in a few places.
___
*[pkg/sql/colexec/hashtable.go, line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20p8GzCwxFZkDigKcF:-M20p8GzCwxFZkDigKcG:b-7jr3nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L304)):*
> ```Go
> 			copy(ht.probeScratch.hashBuffer, ht.probeScratch.next[1:])
> 
> 			for _, hash := range ht.probeScratch.hashBuffer[:batch.Length()] {
> ```

This optimization definitely deserves a comment explaining why we don't need to zero out the whole `ht.probeScratch.first`.
___
*[pkg/sql/colexec/hashtable.go, line 455 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20pNca46PNexfnRVxa:-M20pNca46PNexfnRVxb:budr3z9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L455)):*
> ```Go
> 	ctx context.Context, first, next []uint64, offset, batchSize int,
> ) {
> 	for id := offset + batchSize - 1; id >= offset; id-- {
> ```

This also deserves a comment why we're iterating in reverse order.
___
*[pkg/sql/colexec/hashtable.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20pRsB3SHRy5SfSxOy:-M20pRsB3SHRy5SfSxOz:b-64hfsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L461)):*
> ```Go
> 		hash := next[id]
> 		firstKeyID := first[hash]
> 		if firstKeyID == 0 || uint64(id) < firstKeyID {
> ```

And this `if` as well :)

Also, I think it'd be worth highlighting in the comment to `hashTable.first` that the hash table is built in a such way that `first` always contains `keyID` of the element that was seen first among all that hashed to this hash bucket.

While typing out this comment, I started thinking that this might not work correctly in the case of hash collisions, as you were alluding to earlier, because we're only guaranteed that within a single `next` chain we might have multiple "equality" chains of elements that equal within its chain but are hash collision between "equality" chains, and I fear that we're only keeping the first element of one equality chain that has the lowest `keyID`.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/distinct_test.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20ofxiD43Lk0vNRYMi:-M241HfL5Aady1h7XnDA:b-wn1kjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/distinct_test.go#L133)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

nit: maybe make the order here more randomized? Also I think you should add `nil` in the grouping columns in a few places.
</blockquote></details>

The point of the test was to test when the hash table is performing dedup with various batch size boundaries , the hash table is going to emit the correct tuples. I think randomizing this is going to defeat the purpose of the test since we always "append" the probe table to the build table after deduplication is done on the probe table, at that point build table will only contain tuples with distinct keys the scenario where there exist multiple tuples with same key will never happen. 

Should I add a comment to clarify this or add a `name` field in the test cases to stating the purpose of the test? 
___
*[pkg/sql/colexec/hashtable.go, line 304 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20p8GzCwxFZkDigKcF:-M23tnfX1FmHbfauMuGT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L304)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This optimization definitely deserves a comment explaining why we don't need to zero out the whole `ht.probeScratch.first`.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hashtable.go, line 455 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20pNca46PNexfnRVxa:-M23wgJi74uLXdxQy5E9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L455)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This also deserves a comment why we're iterating in reverse order.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/hashtable.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20pRsB3SHRy5SfSxOy:-M243-Xn8mLwXG4HvKl5:brocj7h) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L461)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

And this `if` as well :)

Also, I think it'd be worth highlighting in the comment to `hashTable.first` that the hash table is built in a such way that `first` always contains `keyID` of the element that was seen first among all that hashed to this hash bucket.

While typing out this comment, I started thinking that this might not work correctly in the case of hash collisions, as you were alluding to earlier, because we're only guaranteed that within a single `next` chain we might have multiple "equality" chains of elements that equal within its chain but are hash collision between "equality" chains, and I fear that we're only keeping the first element of one equality chain that has the lowest `keyID`.
</blockquote></details>

I'm not quite sure if  I understand this. I updated tests to create hash collisions by limiting number of buckets in the hash table and they all seemed to pass. 

I can't really think of a scenario where hash collision is going to cause this to fail. Maybe you can elaborate more on this ?

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 14 of 14 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[pkg/sql/colexec/distinct_test.go, line 133 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20ofxiD43Lk0vNRYMi:-M24PiwP5OpbG76FsaU7:b-3twum2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/distinct_test.go#L133)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

The point of the test was to test when the hash table is performing dedup with various batch size boundaries , the hash table is going to emit the correct tuples. I think randomizing this is going to defeat the purpose of the test since we always "append" the probe table to the build table after deduplication is done on the probe table, at that point build table will only contain tuples with distinct keys the scenario where there exist multiple tuples with same key will never happen. 

Should I add a comment to clarify this or add a `name` field in the test cases to stating the purpose of the test? 


</blockquote></details>

SGTM, adding this explanation as a comment would be helpful.
___
*[pkg/sql/colexec/hashtable.go, line 455 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20pNca46PNexfnRVxa:-M24TXnE54UdRRfq86d9:b4eh1oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L455)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Nice explanation, I see the reasoning now.
___
*[pkg/sql/colexec/hashtable.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20pRsB3SHRy5SfSxOy:-M24Teq9BGXQQ2HZt2-x:b82oq4b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L461)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

I'm not quite sure if  I understand this. I updated tests to create hash collisions by limiting number of buckets in the hash table and they all seemed to pass. 

I can't really think of a scenario where hash collision is going to cause this to fail. Maybe you can elaborate more on this ? 

</blockquote></details>

Yep, makes sense to me now.
___
*[pkg/sql/colexec/hashtable.go, line 465 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TO7L6AmBe7G9AiUU:-M24TO7L6AmBe7G9AiUV:b-m5h4ak) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/colexec/hashtable.go#L465)):*
> ```Go
> 	// `next` is strictly in ascending order. This is crucial to ensure that when
> 	// built in distinct mode, hash table will emit the first distinct tuple it
> 	// encounter. When the next chain is built for build side, this invariant no
> ```

nit: `s/encounter/encounters/g`.
___
*[pkg/sql/colexec/hashtable.go, line 466 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TT3UB38JRyrKgEH2:-M24TT3UB38JRyrKgEH3:bss1xho) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/colexec/hashtable.go#L466)):*
> ```Go
> 	// built in distinct mode, hash table will emit the first distinct tuple it
> 	// encounter. When the next chain is built for build side, this invariant no
> 	// longer hold for the equality chains inside `next`. This is ok however for
> ```

nit: `s/hold/holds/g`.
___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M24TjiL3IvxGqE19jwO:b-nl9qlx) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
> ```Go
> 					args := verifyColOperatorArgs{
> 						anyOrder:       true,
> 						colsForEqCheck: nil,
> ```

I think you should remove `colsForEqCheck` (and the corresponding code in `columnar_utils_test`) entirely because `TestDistinctAgainstProcessor` is the only operator that was using it.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/colexec/hashtable.go, line 461 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M20pRsB3SHRy5SfSxOy:-M24WLKk7vTtYrQztQHn:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9095a540d3e0695dbc1d1f34bc1231d7afd9c9d4/pkg/sql/colexec/hashtable.go#L461)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Yep, makes sense to me now.
</blockquote></details>

Done.
___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M24_KfSCHZkrobuYozN:b10m4ym) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think you should remove `colsForEqCheck` (and the corresponding code in `columnar_utils_test`) entirely because `TestDistinctAgainstProcessor` is the only operator that was using it.
</blockquote></details>

Done.

Also change `anyOrder` to false

**Review by yuzefovich — COMMENTED** [review]
I think you'll need release justification on every commit (and maybe PR description).

> Reviewed 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M24bMDQ6YRFwcH3tSZB:b-wab5wa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M24aaqs9StubxTStjNL:b-yyu2d2) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Done.

Also change `anyOrder` to false 
</blockquote></details>

Hm, I'm not sure that it is correct - in the hash aggregator we can emit tuples out of order, right? We are now only guaranteed that the first encountered tuples among equality chains are emitted, but there is no guarantee on the order between them. I think we want `anyOrder: nOrderedCols < nDistinctCols`.

**Review by Azhng — COMMENTED** [review]
Do we consider this change as low risk ? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M24hIl46IGgYaO03Ljd:b-anwj9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M24eJM-8b64mgUaHmtb:bykw3v7) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, I'm not sure that it is correct - in the hash aggregator we can emit tuples out of order, right? We are now only guaranteed that the first encountered tuples among equality chains are emitted, but there is no guarantee on the order between them. I think we want `anyOrder: nOrderedCols < nDistinctCols`.
</blockquote></details>

We could not guarantee the output order in hash aggregator because it's based on go map, and iterating through go map has no ordering guarantees. 

For unordered distinct case the ordering should be preserved for both within the same equality chain and between different equality chains. (Otherwise the behaviour would be different from row exec unordered distinct). 

So given equality chain `e1` and `e2` let `m1` and `m2` be the first member of the each equality chain. Suppose `e1` and `e2` hashed to different bucket, (since ordering is already guaranteed for elements within a same equality chain) and without loss of generality, suppose `keyID(m1) < keyID(m2)`. Then when we perform deduplicating on the probing side, `m1` will be checked against the equality chain `e1` and `m2` will be checked against equality chain `e2`. Since they are equal, `headID[keyID(m1)] = keyID(m1)` and `headID[keyID(m2)]=keyID(m2)` after the check is complete as the result. In `updateSel` function when we repopulate selection vector, since we loop through `headID` in order from lowest keyID to highest keyID in the batch, and as=lso because `keyID(m1) < keyID(m2)` from above, `m1` is always emitted before `m2` 

And I don't think number of ordered columns has anything to do with the choice of output it emits.

**Review by yuzefovich — COMMENTED** [review]
I think this change qualifies for category 2.

> Reviewed 15 of 15 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M24hvtV80lMSAaGmliP:b-5cf3k6)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M24ho4zDq50esnHTi4H:b-ycvm05) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

We could not guarantee the output order in hash aggregator because it's based on go map, and iterating through go map has no ordering guarantees. 

For unordered distinct case the ordering should be preserved for both within the same equality chain and between different equality chains. (Otherwise the behaviour would be different from row exec unordered distinct). 

So given equality chain `e1` and `e2` let `m1` and `m2` be the first member of the each equality chain. Suppose `e1` and `e2` hashed to different bucket, (since ordering is already guaranteed for elements within a same equality chain) and without loss of generality, suppose `keyID(m1) < keyID(m2)`. Then when we perform deduplicating on the probing side, `m1` will be checked against the equality chain `e1` and `m2` will be checked against equality chain `e2`. Since they are equal, `headID[keyID(m1)] = keyID(m1)` and `headID[keyID(m2)]=keyID(m2)` after the check is complete as the result. In `updateSel` function when we repopulate selection vector, since we loop through `headID` in order from lowest keyID to highest keyID in the batch, and as=lso because `keyID(m1) < keyID(m2)` from above, `m1` is always emitted before `m2` 

And I don't think number of ordered columns has anything to do with the choice of output it emits. 
</blockquote></details>

Ok, that seems to work.

I think it'd be beneficial if you run this test with `nRuns = 300` or something like that, to stress it.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M258KLw2TzkRBlEnWFg:bqcn7ip) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Ok, that seems to work.

I think it'd be beneficial if you run this test with `nRuns = 300` or something like that, to stress it.
</blockquote></details>

It was 0.47s previously with 10 runs. Updated to `nRuns = 150` so the test finishes under 10s.

**Review by yuzefovich — APPROVED** [review]
Nice work! :lgtm: if no failures in 1000 runs.

Also, since we're in stability period, we need a tech lead's approval to merge it (at least previously it was required), so it'd be good to wait for @asubiotto's approval.

> Reviewed 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M259WVO8XCb9KhmLfSP:b-q3a6rz)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M258uM4-zuc3M0bDVVm:b4gqbny) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

It was 0.47s previously with 10 runs. Updated to `nRuns = 150` so the test finishes under 10s. 
</blockquote></details>

I mean we still want to keep the value as `10` runs on master (so that the test time is reasonable), but that you'd update the value to a bigger number manually and run the tests locally. If in 1000 runs there are no failures, then I'd be quite confident in the correctness.

**Review by Azhng — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M25EGCk0MIQv-CazCLz:b-lpawsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I mean we still want to keep the value as `10` runs on master (so that the test time is reasonable), but that you'd update the value to a bigger number manually and run the tests locally. If in 1000 runs there are no failures, then I'd be quite confident in the correctness.
</blockquote></details>

Sounds good. Ran it for 10000 runs, no errors :stuck_out_tongue: 

https://gist.github.com/Azhng/1a93c7dfbbb9bf277dc5e372a77d9257

**Review by yuzefovich — COMMENTED** [review]
> Reviewed 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/distsql/columnar_operators_test.go, line 243 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-M24TjiL3IvxGqE19jwN:-M25Eixs8H6GAfJAZPl7:bjwa7hh) ([raw file](https://github.com/cockroachdb/cockroach/blob/83ffed32b057e5b77726fcd423e04dc15f151691/pkg/sql/distsql/columnar_operators_test.go#L243)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Sounds good. Ran it for 10000 runs, no errors :stuck_out_tongue: 

https://gist.github.com/Azhng/1a93c7dfbbb9bf277dc5e372a77d9257
</blockquote></details>

:tada:

**Comment by asubiotto** [comment]
I'm still a bit uncomfortable about enabling this by default. @Azhng, could you hold off merging this until we discuss during the standup?

**Comment by asubiotto** [comment]
We discussed and we won't be enabling this in `vectorize=auto` to reduce the risk profile of vectorized execution in 20.1.

**Comment by yuzefovich** [comment]
:lgtm:

**Review by yuzefovich — APPROVED** [review]
> Reviewed 11 of 11 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906)* status: :shipit: complete! 1 of 0 LGTMs obtained

**Review by Azhng — COMMENTED** [review]
@asubiotto  sorry I forgot to update the commits and the commit messages. CTAL ? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M2K6s8N--bZtr-HJhD9:b5dq967)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by asubiotto** [comment]
I think in the commit message it might be useful to state why this change is low risk. The reason here is that we're not turning it on by default.

**Review by Azhng — COMMENTED** [review]
@asubiotto PTAL 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M2dHeo89C-rUahHT0Va:b-fq2gv4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by asubiotto — APPROVED** [review]
:lgtm: sorry I let this slip!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45906#-:-M2dRMRrC4ougZhbdAyi:b-8nd5f7)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

**Comment by Azhng** [comment]
TFTR!

bors r+

**Pull request merged**
